### ISSUE TITLE
Opt out support for S3 bucket versioning via stack parameters

### ISSUE THREAD

**Issue by ruffsl** [runs-on/runs-on#191]
Not sure extending such an opt out to all other S3 buckets used internally for the RunsOn cloud formation would be need here, as perhaps RunsOn could grow to rely on that functionally in the future, but for the user facing cache bucket(s) (or at least the `/cache` sub folder - but I think versioning is bucket wide), it would be nice to be able to opt out of versioning in order to save on storage/retention resource usage, particularly for cache overwrite operations.

Unlike GitHub's stock `action/cache` action, the `runs-on/cache` action does enable the use of mutable cache keys, being that the backend is a simple namespaced hierarchy of objects in conventional S3 bucket. This allows for overwriting cache blobs without the need for alternately incrementing cache key via postfix or more complex retention/cleanup operations. This is valuable when updating Gigabyte sized cache blob stores, like from CCache for C/C++ compilers or incremental workspace builds, without duplicate archiving old cache keys of basically the same thing, where versioning offers little utility but can incur large costs.

**Comment by ruffsl** [comment]
One implementation/migration detail may be the fact that switching to a versioned bucket is permanent: 

> You enable and suspend versioning at the bucket level. After you version-enable a bucket, it can never return to an unversioned state. But you can suspend versioning on that bucket.
> https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html

So I surmise that such an opt out could only suspended versioning (aside from simply removing and recreating the cache bucket, but probably undesirable for migration/upgrades), so probably want to keep the `NoncurrentVersionExpiration` setting in the bucket policies regardless of status of `VersioningConfiguration` to ensure concurrent version are still clean up after opt-out/suspension?

- https://docs.aws.amazon.com/AmazonS3/latest/userguide/AddingObjectstoVersionSuspendedBuckets.html

**Comment by crohr** [comment]
@ruffsl good catch, in that case the best course of action would be to manually suspend versioning for your own usage. I don't think a lot of users are overriding cache keys with runs-on/cache. Would that be acceptable?

**Comment by ruffsl** [comment]
We could just manually suspend it in the S3 policy for the cache bucket, but wouldn't there be a risk that an upgrade later on would un-suspended it if said policy was ever auto updated by the stack? In that case, it would be nice to have a stack parameter to persistent this degree of config, regardless of who in the team remembers to upgrade the cloudformation - As I've had to re-patch the same S3 cache bucket policy in the past post-upgrade to restore developer read access, so our team could continue to download CI cache assets for regular inspection and bootstrapping local caches.

**Comment by crohr** [comment]
In #219 I'm going to simply suspend versioning on the cache bucket for now. It was added for compliance reason, but it just doesn't make sense. And I would prefer to avoid adding another stack parameter just for that.

**Comment by ruffsl** [comment]
> In #219 I'm going to simply suspend versioning on the cache bucket for now.

Awesome, will jump to that when released then.

> It was added for compliance reason, but it just doesn't make sense.

 Yeah, not much point in re-visioning a cache bucket, just adds to the cost. I'm guessing for others who rolled this out, their bucket versioned artifacts will just age out like anything else from the default lifecycle policy?

**Comment by crohr** [comment]
> Yeah, not much point in re-visioning a cache bucket, just adds to the cost. I'm guessing for others who rolled this out, their bucket versioned artifacts will just age out like anything else from the default lifecycle policy?

Yes!
